### ISSUE TITLE
MATERIALIZE PROJECTION Make ClickHouse Daemon Exits Follow up

### ISSUE THREAD

**Issue by ggchangan** [ClickHouse/ClickHouse#30781]
I have created an issue for this bug, which is https://github.com/ClickHouse/ClickHouse/issues/27512. And there is a pr for this bug, Thanks a lot. 
However, I test it on 21.8.8.29 and 21.10.2.15. The ClickHouse Daemon Exists issue still exists.

**Comment by ggchangan** [comment]
After several tests, I still haven't found how to produce this issue in small dataset and now I try to read your source code, but it's so difficult for me. However I get the following call stack when this issue happens, whether this stack help find this issue?

```
/NOT_FOUND_COLUMN_IN_BLOCK
2021.11.15 04:34:47.079884 [ 129 ] {} <Error> test.f_test_local (cc7f7aed-d78d-4a81-8c7f-7aedd78dda81): auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(DB::StorageReplicatedMergeTree::LogEntryPtr &) const: Code: 10. DB::Exception: Not found column columnError in block. There are only columns: columnA, columnA, columnA, ^@^A^A^A^@^A^A^@^A^@^A^@^A^@^A^@^A^A^A^A^@^A^A^A^A^A^@^A^A, columnA, columnB, columnC, columnD, columnE, if(equals(columnF, 'TIMEOUT'), columnG, 0), if(equals(error_code, 'HTTP_ERROR'), columnG, 0), if(notIn(error_code, ('HTTP_ERROR', 'TIMEOUT')), columnG, 0), if(equals(error_code, 'NO_BIDS'), columnG, 0), if(equals(error_code, 'MALFORMED_RESPONSE'), columnG, 0), if(equals(error_code, 'BID_RESPONSE_ID_NOMATCH'), columnG, 0), if(and(greater(bitAnd(columnA, 4), 0), equals(error_code, '')), columnG, 0), if(equals(error_code, 'TIMEOUT'), columnH, 0), if(in(error_code, ('HTTP_ERROR', 'MALFORMED_RESPONSE')), columnH, 0). (NOT_FOUND_COLUMN_IN_BLOCK), Stack trace (when copying this message, always include the lines below):
0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x944bdda in /usr/bin/clickhouse
1. DB::Block::getPositionByName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x10440fa6 in /usr/bin/clickhouse
2. DB::InterpreterSelectQuery::executeAggregation(DB::QueryPlan&, std::__1::shared_ptr<DB::ActionsDAG> const&, bool, bool, std::__1::shared_ptr<DB::InputOrderInfo const>) @ 0x10caa5ba in /usr/bin/clickhouse
3. DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::shared_ptr<DB::IBlockInputStream> const&, std::__1::optional<DB::Pipe>) @ 0x10ca1167 in /usr/bin/clickhouse
4. DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0x10c9f480 in /usr/bin/clickhouse
5. DB::InterpreterSelectQuery::execute() @ 0x10ca2951 in /usr/bin/clickhouse
6. DB::ProjectionDescription::calculate(DB::Block const&, std::__1::shared_ptr<DB::Context const>) const @ 0x1125e9d7 in /usr/bin/clickhouse
7. DB::MergeTreeDataMergerMutator::writeWithProjections(std::__1::shared_ptr<DB::IMergeTreeDataPart>, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::vector<DB::ProjectionDescription const*, std::__1::allocator<DB::ProjectionDescription const*> > const&, std::__1::shared_ptr<DB::IBlockInputStream>, DB::IMergedBlockOutputStream&, long, DB::MergeListElement&, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&, std::__1::shared_ptr<DB::Context const>, DB::IMergeTreeDataPart::MinMaxIndex*) @ 0x1169b2de in /usr/bin/clickhouse
8. DB::MergeTreeDataMergerMutator::mutateSomePartColumns(std::__1::shared_ptr<DB::IMergeTreeDataPart const> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::set<std::__1::shared_ptr<DB::IMergeTreeIndex const>, std::__1::less<std::__1::shared_ptr<DB::IMergeTreeIndex const> >, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeIndex const> > > const&, std::__1::set<DB::ProjectionDescription const*, std::__1::less<DB::ProjectionDescription const*>, std::__1::allocator<DB::ProjectionDescription const*> > const&, DB::Block const&, std::__1::shared_ptr<DB::IMergeTreeDataPart>, std::__1::shared_ptr<DB::IBlockInputStream>, long, std::__1::shared_ptr<DB::ICompressionCodec> const&, DB::MergeListElement&, DB::ExecuteTTLType, bool, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&, std::__1::shared_ptr<DB::Context const>) @ 0x116981d6 in /usr/bin/clickhouse
9. DB::MergeTreeDataMergerMutator::mutatePartToTemporaryPart(DB::FutureMergedMutatedPart const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::MutationCommands const&, DB::MergeListElement&, long, std::__1::shared_ptr<DB::Context const>, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&) @ 0x1168fc5c in /usr/bin/clickhouse
10. DB::StorageReplicatedMergeTree::tryExecutePartMutation(DB::ReplicatedMergeTreeLogEntry const&) @ 0x113730e0 in /usr/bin/clickhouse
11. DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x1135df5c in /usr/bin/clickhouse
12. ? @ 0x11424f3f in /usr/bin/clickhouse
13. DB::ReplicatedMergeTreeQueue::processEntry(std::__1::function<std::__1::shared_ptr<zkutil::ZooKeeper> ()>, std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::__1::function<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>) @ 0x117f312c in /usr/bin/clickhouse
14. DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x113ac63d in /usr/bin/clickhouse
15. ? @ 0x11586877 in /usr/bin/clickhouse
16. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x948fb98 in /usr/bin/clickhouse
17. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0x949173f in /usr/bin/clickhouse
18. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x948ce7f in /usr/bin/clickhouse
19. ? @ 0x9490763 in /usr/bin/clickhouse
20. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
21. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
 (version 21.10.2.15 (official build))
```

**Comment by amosbird** [comment]
> @amosbird I test it on 21.11.3.6, it works. However I still don't know why it doesn't work on 21.10.2.15.

It means the issue has been fixed in recent PRs)

**Comment by ggchangan** [comment]
@amosbird https://github.com/ClickHouse/ClickHouse/pull/29337 has been merged to 21.10.3.9. However, this issue still happens, is there any PRs which are related to this issue are not merged to 21.10?

**Comment by amosbird** [comment]
I cannot tell without a reproduceable case. Please try to construct one using existing dataset (you can utilize `clickhouse-obfuscator`)
